File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        def util(i, r):
            if i == n:
                res.append(r[:])
                return
            else:
                r.append(nums[i])
                util(i+1, r)
                r.pop()
                util(i+1, r)
        util(0, [])
        return res

File: Data Structures & Algorithms/subsets-ii/submission-1.cpp
class Solution {
public:

    void util(vector<vector<int>>& res, vector<int>& nums, vector<int>& curr, int i, int &n) {
        res.push_back(curr);
        for (int j = i; j < n; j++) {
            if (j > i and nums[j] == nums[j-1]) continue;
            curr.push_back(nums[j]);
            util(res, nums, curr, j+1, n);
            curr.pop_back();
        }
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> curr;
        int n = nums.size();
        util(res, nums, curr, 0, n);
        return res;
    }
};